// payumoney-pay/index.js
var request = require('request');
var payment = require('./pay_helper');

var API_ENDPOINT='https://sandboxsecure.payu.in/';
var Payment = {
    //SETUP AUTH DATA
    AUTH_SETUP:{
        key:'YOUR KEY HERE',
        salt:'YOUR SALT HERE',
        auth_header:'YOUR AUTHORIZATION HEADER HERE',
        url:'YOUR URL HERE',
        test_mode:'https://sandboxsecure.payu.in/',
        prod_mode:'https://secure.payu.in/',
        headers:'HEADERS HERE',
        surl : 'https://success-url.in',
        furl: 'https://fail-url.in'
    },
    //ACCESSING REST API 
    REST_API:{
            prod_mode: "https://www.payumoney.com/",
            test_mode: 'https://test.payumoney.com/',  
    },
    //API URL
    API_URL:{
        CRT:'_payment',
        ASK:'payment/op/getPaymentResponse?',
        CHK:'payment/payment/chkMerchantTxnStatus?',
        SMS:'payment/payment/smsInvoice?'
    },
    //SET AUTH DATA
    setAuthData :function (payumode,key,salt,auth_key,surl,furl) {
        this.AUTH_SETUP.key=key;//SET KEY
        this.AUTH_SETUP.salt=salt;//SET SALT
        this.AUTH_SETUP.auth_header=auth_key;//SET AUTH HEADER
        this.AUTH_SETUP.headers= {
            'Authorization': this.AUTH_SETUP.auth_header,
            'Content-Type': 'application/json'
        }//SET HEADER
        this.AUTH_SETUP.surl=surl,
        this.AUTH_SETUP.furl=furl,
        this.AUTH_SETUP.mode=payumode
    },
    //SET URL
    setUrl: function (mode,isRestAPI) {
        if(mode) {
            if(isRestAPI)
                this.API_ENDPOINT=this.REST_API.prod_mode;
            else
                this.API_ENDPOINT=this.AUTH_SETUP.prod_mode;
        }
        else {
            if(isRestAPI)
                this.API_ENDPOINT=this.REST_API.test_mode;
            else
                this.API_ENDPOINT=this.AUTH_SETUP.test_mode;
        }
    },
   //GET OPTIONS AND HEADERS
    getOption:function (rest_url,data,action,isRestAPI) {
        this.setUrl(this.AUTH_SETUP.mode,isRestAPI);//SET URL
        let options;
        if(action=="CRT"){
            options = {
                url: this.API_ENDPOINT+rest_url,
                form:payment.setHashData(data,this.AUTH_SETUP.key,this.AUTH_SETUP.salt,this.AUTH_SETUP.surl,this.AUTH_SETUP.furl),
                headers:this.AUTH_SETUP.headers
            };
        }
        if (action=="RFND") {
            options = {
                url: this.API_ENDPOINT+rest_url,
            };
        }
        if (action=="ASK") {

            options = {
                url: this.API_ENDPOINT+rest_url+'merchantKey='+this.AUTH_SETUP.key+'&merchantTransactionIds='+data,
                headers:this.AUTH_SETUP.headers
            };
        }
        if (action=="CHK") {

            options = {
                url: this.API_ENDPOINT+rest_url+'merchantKey='+this.AUTH_SETUP.key+'&merchantTransactionIds='+data,
                headers:this.AUTH_SETUP.headers
            };
        }
        if (action=="SMS") {

            options = {
                url: this.API_ENDPOINT+rest_url+'customerName='+data.customerName+'&customerMobileNumber='+data.customerMobileNumber+'&description='+data.description+'&amount='+data.amount,
                headers:this.AUTH_SETUP.headers
            };
        }
        return options;
    },
    //MAKE PAYMENT
    makePayment:function (data,callback) {
        request.post(this.getOption(this.API_URL.CRT,data,"CRT",false),function (error, response, body) {
            if(error) throw error
            else {
                var result = response;
                callback(error, result);
            }
          });
    },
    getPaymentResponse : function (tnxId,callback) {
        request.post(this.getOption(this.API_URL.ASK,tnxId,"ASK",true),function (error, response, body) {
            if(error) throw error
            else {
                var result = response
                callback(error, result);
            }
          });
    },
    checkPaymentResponse : function (tnxId,callback) {
        request.post(this.getOption(this.API_URL.CHK,tnxId,"CHK",true),function (error, response, body) {
            if(error) throw error
            else {
                var result = response
                callback(error, result);
            }
          });
    },
    sendSMSInvoice : function (data,callback) {
        request.post(this.getOption(this.API_URL.SMS,data,"SMS",true),function (error, response, body) {
            if(error) throw error
            else {
                var result = response
                callback(error, result);
            }
          });
    }
}
module.exports = Payment;
