// payumoney-pay/pay_helper.js
var sha512 = require('js-sha512');

var Payment= {
    setHashData:function(RequestBody,auth_key,auth_salt,surl,furl) {
        RequestData = {
            key:auth_key,
            txnid:RequestBody.txnid,
            amount: RequestBody.amount,
            firstname:RequestBody.firstname,
            email: RequestBody.email,
            phone: RequestBody.phone,
            productinfo: RequestBody.productinfo,
            salt:auth_salt
        }
        var hashSequence= RequestData.key+ '|' + RequestData.txnid + '|' + RequestData.amount+ '|' +RequestData.productinfo + '|' +RequestData.firstname + '|' + RequestData.email + '|||||||||||'+RequestData.salt+'';
        var hash=sha512(hashSequence);
        var RequestHashData = {
            key:auth_key,
            hash:hash,
            txnid: RequestData.txnid,
            amount: RequestData.amount,
            firstname:RequestData.firstname,
            email: RequestData.email,
            phone: RequestData.phone,
            productinfo: RequestData.productinfo,
            service_provider:'payu_paisa',
            surl : surl,
            furl: furl
        }

        return RequestHashData;
    },
}
module.exports=Payment;
